Mathlib/Data/QuotLike.lean

```diff
@@ -0,0 +1,666 @@
+* `mkQ_h` `⟦a⟧_h`       : the canonical quotient map inferred from
+                            the hint via typeclass `QuotLike.Hint`
+                            or the hint and the input type via typeclass `QuotLike.HasQuotHint`
+-/
+
+/--
+The definition of `QuotLike`. It is needed to avoid name clashes
+between `QuotLike.rec` `QuotLike.recOn` and the autogenerated recursors.
+-/
+class QuotLikeStruct (Q : Sort*) (α : outParam Sort*) (r : outParam (α → α → Prop)) where
+  /-- The canonical quotient map. Inferred from the output type. -/
+  mkQ : α → Q := by exact Quot.mk _
+  /-- The canonical map from quotient to `Quot r`. -/
+  toQuot : Q → Quot r := by exact (·)
+  toQuot_mkQ : ∀ a, toQuot (mkQ a) = Quot.mk r a := by exact fun _ ↦ rfl
+  mkQ_surjective : Function.Surjective mkQ := by exact Quot.exists_rep
+  /--
+  The analogue of `Quot.sound`: If `a` and `b` are related by the relation,
+  then they have equal equivalence classes.
+  -/
+  sound {a b : α} : r a b → mkQ a = mkQ b := by exact Quot.sound
+
+/--
+The class `QuotLike Q α r` expresses the type `Q` is canonically isomorphic to `Quot α r`.
+
+Using `QuotLike` API for type synonyms of `Quot` and `Quotient` will avoid defeq abuse caused by
+directly using `Quot` and `Quotient` APIs.
+-/
+abbrev QuotLike := QuotLikeStruct
+
+export QuotLikeStruct (mkQ)
+
+namespace QuotLike
+
+export QuotLikeStruct (mkQ toQuot)
+
+alias toQuot_mkQ := QuotLikeStruct.toQuot_mkQ
+alias sound := QuotLikeStruct.sound
+
+attribute [simp] toQuot_mkQ
+
+open Lean Elab Term Meta Qq
+
+@[inherit_doc mkQ]
+elab "⟦" a:term "⟧" : term <= Q => do
+  synthesizeSyntheticMVars
+  let Q ← instantiateMVars Q
+  if Q.isMVar then
+    tryPostpone
+    throwError "The output type must be known."
+  let v ← match ← inferType Q with | .sort v => pure v | _ => mkFreshLevelMVar
+  have Q : Q(Sort v) := Q
+  let α ← mkFreshExprMVarQ q(Sort $(← mkFreshLevelMVar))
+  let r ← mkFreshExprMVarQ q($α → $α → Prop)
+  let .some inst ← trySynthInstanceQ q(QuotLike $Q $α $r) |
+    tryPostpone
+    throwError "Cannot find `QuotLike` instance for type `{Q}`."
+  pure q(@mkQ $Q $α $r $inst $(← Qq.elabTermEnsuringTypeQ a q($α)))
+
+open PrettyPrinter.Delaborator SubExpr in
+/-- Delaborator for `mkQ` -/
+@[delab app.QuotLikeStruct.mkQ]
+def delabMkQ : Delab := do
+  guard <| (← getExpr).isAppOfArity' ``mkQ 5
+  let a ← withNaryArg 4 delab
+  `(⟦$a⟧)
+
+@[inherit_doc mkQ]
+macro "⟦" a:term " : " α:term "⟧" : term => `(⟦($a : $α)⟧)
+
+/--
+`QuotLike.HasQuot` is used for deriving `QuotLike` instances from the input type.
+
+The instances of `QuotLike.HasQuot` may be defined for types that have specific uses.
+
+```
+instance : QuotLike.HasQuot ZFSet PSet PSet.Equiv where
+```
+
+They are also usually defined as scoped or local instances.
+
+```
+scoped instance {α} [s : Setoid α] : QuotLike.HasQuot (Quotient s) α (· ≈ ·) where
+```
+-/
+class HasQuot (Q : outParam Sort*) (α : Sort*) (r : outParam (α → α → Prop))
```

Comment:
Why do we need this, as well as `QuotLike`? And why do we need both the `⟦x⟧` notation as well as the `mkQ' x` notation?